### PR TITLE
Create a raw node instead of a file node when there is no content.

### LINKED ISSUE #4688: ipfs add --raw-leaves return protobuf instead or raw on zero length file
#### Version information:
go-ipfs version: 0.4.13-
Repo version: 6
System version: amd64/windows
Golang version: go1.9.2

#### Description:
```
>ipfs add -LD --raw-leaves zero-len-file
2018-02-12 21:53:53.609375 INFO cmd/ipfs main.go:301: IPFS_PATH .ipfs
2018-02-12 21:53:53.609375 DEBUG cmd/ipfs main.go:295: calling pre-command hooks...
2018-02-12 21:53:53.625000 DEBUG cmd/ipfs main.go:342: executing command locally
2018-02-12 21:53:53.625000 DEBUG core/commands add.go:157: Total size of file being added: 0
2018-02-12 21:53:53.656250 DEBUG blockservice blockservice.go:160: BlockService GetBlock: 'QmS7RSorsfu1MJpBjcu7L1W2QfsoFDKSrvqAiHzLY4EUH4'
2018-02-12 21:53:53.656250 DEBUG blockservice blockservice.go:160: BlockService GetBlock: 'QmWWC1GHCa7at4K6oMK7Wm7odYCphhRvXYXKAJBreMhRLu'
2018-02-12 21:53:53.656250 DEBUG blockservice blockservice.go:160: BlockService GetBlock: 'QmSKboVigcD3AY4kLsob117KJcMHvMUu6vNFqk1PQzYUpp'
2018-02-12 21:53:53.671875 DEBUG blockservice blockservice.go:160: BlockService GetBlock: 'QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn'
2018-02-12 21:53:53.671875 DEBUG core/commands add.go:397: add progress: zero-len-file 0
 0 B / ? [--------------------------------------------------------------------------------------------------------------------------------
added QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH zero-len-file
2018-02-12 21:53:53.671875 DEBUG blockservice blockservice.go:160: BlockService GetBlock: 'QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH'
2018-02-12 21:53:53.671875 DEBUG blockservice blockservice.go:160: BlockService GetBlock: 'QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH'
2018-02-12 21:53:53.687500 INFO cmd/ipfs main.go:240: Shutting down node...
2018-02-12 21:53:53.687500 DEBUG core core.go:531: core is shutting down...
2018-02-12 21:53:53.687500 DEBUG blockservice blockservice.go:251: blockservice is shutting down...
```

Result: QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH
```
ipfs dag get QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH
{"data":"CAIYAA==","links":[]}
```

Real zero length raw block: zb2rhmy65F3REf8SZp7De11gxtECBGgUKaLdiDj7MCGCHxbDW
```
>ipfs block put -LD -f"raw" zero-len-file
2018-02-12 22:09:45.250000 INFO cmd/ipfs main.go:301: IPFS_PATH .ipfs
2018-02-12 22:09:45.250000 DEBUG cmd/ipfs main.go:295: calling pre-command hooks...
2018-02-12 22:09:45.250000 DEBUG cmd/ipfs main.go:342: executing command locally
2018-02-12 22:09:45.296875 DEBUG blockservice blockservice.go:160: BlockService GetBlock: 'QmS7RSorsfu1MJpBjcu7L1W2QfsoFDKSrvqAiHzLY4EUH4'
2018-02-12 22:09:45.296875 DEBUG blockservice blockservice.go:160: BlockService GetBlock: 'QmWWC1GHCa7at4K6oMK7Wm7odYCphhRvXYXKAJBreMhRLu'
2018-02-12 22:09:45.296875 DEBUG blockservice blockservice.go:160: BlockService GetBlock: 'QmSKboVigcD3AY4kLsob117KJcMHvMUu6vNFqk1PQzYUpp'
2018-02-12 22:09:45.296875 DEBUG blockservice blockservice.go:160: BlockService GetBlock: 'QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn'
2018-02-12 22:09:45.312500 DEBUG core/commands block.go:188: BlockPut key: '"zb2rhmy65F3REf8SZp7De11gxtECBGgUKaLdiDj7MCGCHxbDW"'
zb2rhmy65F3REf8SZp7De11gxtECBGgUKaLdiDj7MCGCHxbDW
2018-02-12 22:09:45.312500 INFO cmd/ipfs main.go:240: Shutting down node...
2018-02-12 22:09:45.312500 DEBUG core core.go:531: core is shutting down...
2018-02-12 22:09:45.312500 DEBUG blockservice blockservice.go:251: blockservice is shutting down...
```

### PR DESCRIPTION
**Pull request by kevina**
This fixes things so when raw-leaves are enabled a zero size file creates a zero size raw leaf.  ~~When raw-leaves are not enabled the hash created changes from `QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH` to `Qmdsf68UUYTSSx3i4GtDJfxzpAEZt7Mp23m3qa36LYMSiW`, since the type field
changed from `TFile` to `TRaw`.~~

Closes #4688

### CODE REVIEW

File: importer/helpers/dagbuilder.go [hunk 1]
```diff
@@ -122,6 +122,34 @@ func (db *DagBuilderHelper) NewUnixfsNode() *UnixfsNode {
 	return n
 }
 
+// NewUnixfsLeaf creates a lead node filled with date
```
Comment (hsanjuan) [inline importer/helpers/dagbuilder.go]: This is going to cause a golint warning. (and `date` should be `data`)

**Comment by kevina** [comment]
Jenkins is complianing about this:
```
Following Go code is not formatted.
-----------------------------------
./fuse/ipns/ipns_test.go
-----------------------------------
Run 'go fmt ./...' in your source directory
mk/golang.mk:48: recipe for target 'test_go_fmt' failed
```
Except (1) I did not touch that file (2) that file is already go formatted in that commit as `gofmt -d ./fuse/ipns/ipns_test.go` returns nothing and (3) when I ran `bin/test-go-fmt` it doesn't report anything.   The script itself is very simple so I am not sure what the problem is.

**Comment by magik6k** [comment]
This is caused by https://github.com/ipfs/go-ipfs/commit/682ca11dd8fbd63e447e9a0290463b2bba208f50, which got merged to master in https://github.com/ipfs/go-ipfs/pull/4695

**Comment by hsanjuan** [comment]
hmm, my bad, I will PR that. it seems it's the order of the imports.

**Comment by hsanjuan** [comment]
#4702

**Comment by Kubuxu** [comment]
@whyrusleeping are you ok with changing hash of empty file?

**Comment by Stebalien** [comment]
> @whyrusleeping are you ok with changing hash of empty file?

Doesn't this only apply when raw leaves (experimental) are enabled?

**Comment by magik6k** [comment]
Just tried, for me it changes to `zb2rhmy65F3REf8SZp7De11gxtECBGgUKaLdiDj7MCGCHxbDW` with raw-leaves and to `Qmdsf68UUYTSSx3i4GtDJfxzpAEZt7Mp23m3qa36LYMSiW` without (it's `QmbFMke1KXqnYyBBWxB74N4c5SBnJMVAiMNRcGu6x1AwQH` in both cases on master now)

**Comment by kevina** [comment]
@Kubuxu I added a special case to avoid changing the hash, I am removing the blocked label.

**Comment by kevina** [comment]
@whyrusleeping @Kubuxu @Stebalien I think this should be good to go now.

**Comment by schomatis** [comment]
Should the trickle builder also be taken into consideration here?

**Comment by kevina** [comment]
> Should the trickle builder also be taken into consideration here?

I don't think so.  The trickle dag format never returns a leaf node's hash, so it will make sense that an empty node is simply a node with no children, rather than an empty leaf node.

**Pull request merged**